# nf-core/yascp: Output

## Introduction

This guide provides a comprehensive overview of the outputs generated by the pipeline.

The pipeline will create the following files in your working directory:

```console
work            # Directory containing the nextflow working files
results         # Finished results (configurable, see below)
.nextflow_log   # Log file from Nextflow
# Other nextflow hidden files, eg. history of pipeline runs and old logs.
```

## Pipeline Overview

Utilizing [Nextflow](https://www.nextflow.io/), our pipeline orchestrates a series of data processing steps. The structure of the overall results folder is outlined below, offering a snapshot of the diverse outputs from different stages of the pipeline:

- **[preprocessing](#preprocessing)**: ???
- **[doublet_detection](#doublet_detection)**: A folder with identified artificial doublet cells.
- **[deconvolution](#deconvolution)**: A folder with the results of disentangling mixtures of cells from different donors.
- **[celltype_assignemt](#celltype_assignemt)**: A folder with cells classified into types.
- **[clustering_and_integration](#clustering_and_integration)**:  A folder with cells grouped based on similarities and integrating datasets.
- **[citeseq](#citeseq)**: A folder with CITE-seq (Cellular Indexing of Transcriptomes and Epitopes by Sequencing) data.
- **[handover](#handover)**: ???
- **[pipeline_info]**: ???
- **[yascp_inputs]**: ???

Detailed explanations of each output folder and their corresponding steps are provided below:

## preprocessing
```
preprocessing/
├── data_modalities_split
│   ├── filterd
│   │   └── Pool1
│   │       ├── Gene_Expression-Pool1.h5ad
│   │       └── Pool1__Gene_Expression
│   ├── preprocess
│   │   └── Pool1
│   │       ├── Gene_Expression-Pool1.h5ad
│   │       └── Pool1__Gene_Expression
│   └── raw
│       └── Pool1
│           ├── Gene_Expression-Pool1.h5ad
│           └── Pool1__Gene_Expression
├── recourses
│   ├── Done.tmp
│   ├── full_test_dataset
│   ├── input_test_data_file.tsv
│   └── input_test_vcf_file.tsv
└── subset_genotypes
    ├── Genotype___AllExpectedGT_Pool1
    └── Genotypes_all_pools.tsv

```

The folder preprocessing contains the next folders
- **[data_modalities_split]  raw and filtered quantification data devided by the modality (citeseq or hashtag)
- **[recourses] genotype assembly and and input files for test dataset
- **[subset_genotypes] contains divided genotipes from pools. Apears only if you specify vcf ids in the input tsv file (improve writing) 

## doublet_detection
```
doublet_detection
├── DoubletDecon
│   └── Pool1__DoubletDecon_doublets_singlets.tsv
├── DoubletFinder
│   └── Pool1__DoubletFinder_doublets_singlets.tsv
├── doublet_results_combined
│   └── Pool1__doublet_results_combined.tsv
├── droplet_type_distribution
│   └── Pool1__droplet_type_distribution.png
├── scDblFinder
│   └── Pool1__scDblFinder_doublets_singlets.tsv
├── SCDS
│   └── Pool1__scds_doublets_singlets.tsv
└── scrublet
    ├── plots
    │   ├── Pool1boxplot_total_umi_counts.png
    │   ├── Pool1histogram_multiplet_scores_log.png
    │   ├── Pool1histogram_multiplet_scores.png
    │   └── Pool1histogram_multiplet_zscores.png
    └── Pool1scrublet.tsv
```
DoubletFinder, DoubletDecon, scDblFinder, SCDS contain tsv files per pool with a barcode and label of whether it's a singlet or a doublet
scrublet contains tsv files per pool with a barcode and label of whether it's a multiplet or not and a folder with plots
doublet_results_combined contains tsv files per pool with barcode and labels from scrublet,	scds, scDblFinder, DoubletDecon, DoubletFinder
droplet_type_distribution contains png files with graphs showing droplet-type distribution

## deconvolution

- **[vireo_raw](#vireo_raw)**:
- **[vireo_processed](#vireo_processed)**:
- **[vireo_sub](#vireo_sub)**:
- **[infered_genotypes](#infered_genotypes)**:
- **[split_donor_h5ad](#split_donor_h5ad)**:
- **[filepaths](#filepaths)**:
- **[cellsnp](#cellsnp)**:
- **[existing_cellsnp](#existing_cellsnp)**: (name can be different)
- **[concordances](#concordances)**:
- **[gtmatch](#gtmatch)**:
infered_genotypes



## vireo_raw
```
vireo_raw
├── correlations.png
├── Pool1
│   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
│   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
│   ├── sub_Pool1_Expected.vcf.gz
│   └── vireo_Pool1
├── donor_corelations_matrix.tsv
└── matched_donors.txt

```
contains genotypes if run in genotype-aware mode

## vireo_processed
```
vireo_processed
├── assignments_all_pools.tsv
└── Pool1
    ├── GT_replace_donor_ids_false.tsv
    ├── GT_replace_GT_donors.vireo_false.vcf.gz
    ├── GT_replace_Pool1_assignments_false.tsv
    ├── GT_replace_Pool1__exp.sample_summary_false.txt
    └── GT_replace_Pool1.sample_summary_false.txt
```
genotypes renamed to imitate genotype absent mode to ensure the consistency in the downstream pipeline tasks

## vireo_sub
```
vireo_sub
├── Pool1
│   ├── vireo_____1
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
│   │   ├── sub_Pool1_Expected.vcf.gz
│   │   └── vireo_Pool1___1
│   ├── vireo_____10
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
│   │   ├── sub_Pool1_Expected.vcf.gz
│   │   └── vireo_Pool1___10
│   ├── vireo_____2
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
│   │   ├── sub_Pool1_Expected.vcf.gz
│   │   └── vireo_Pool1___2
│   ├── vireo_____3
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
│   │   ├── sub_Pool1_Expected.vcf.gz
│   │   └── vireo_Pool1___3
│   ├── vireo_____4
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
│   │   ├── sub_Pool1_Expected.vcf.gz
│   │   └── vireo_Pool1___4
│   ├── vireo_____5
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
│   │   ├── sub_Pool1_Expected.vcf.gz
│   │   └── vireo_Pool1___5
│   ├── vireo_____6
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
│   │   ├── sub_Pool1_Expected.vcf.gz
│   │   └── vireo_Pool1___6
│   ├── vireo_____7
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
│   │   ├── sub_Pool1_Expected.vcf.gz
│   │   └── vireo_Pool1___7
│   ├── vireo_____8
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
│   │   ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
│   │   ├── sub_Pool1_Expected.vcf.gz
│   │   └── vireo_Pool1___8
│   └── vireo_____9
│       ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz
│       ├── dubs_removed__Study_Merge_AllExpectedGT_QW4IKXM1N_out.vcf.gz.csi
│       ├── sub_Pool1_Expected.vcf.gz
│       └── vireo_Pool1___9
```
vireo permutation to ensure stability in cell assignment


## infered_genotypes
```
infered_genotypes
├── Pool1
│   ├── Pool1_headfix_vireo.vcf.gz
│   └── Pool1_headfix_vireo.vcf.gz.tbi
```
genotypes called from single-cell data per (per donor in a pool)

## split_donor_h5ad
```
split_donor_h5ad
└── Pool1
    ├── cell_belongings.tsv
    ├── donor_level_anndata
    │   ├── donor0.Pool1.barcodes.tsv
    │   ├── donor0.Pool1.h5ad
    │   ├── donor1.Pool1.barcodes.tsv
    │   ├── donor1.Pool1.h5ad
    │   ├── donor2.Pool1.barcodes.tsv
    │   ├── donor2.Pool1.h5ad
    │   ├── doublet.Pool1.barcodes.tsv
    │   ├── doublet.Pool1.h5ad
    │   ├── unassigned.Pool1.barcodes.tsv
    │   └── unassigned.Pool1.h5ad
    ├── Pool1.donors.h5ad.assigned.tsv
    ├── Pool1__donors.h5ad.assigned.tsv
    ├── Pool1.donors.h5ad.tsv
    ├── Pool1__donors.h5ad.tsv
    ├── Pool1_exp__donor_n_cells.tsv
    ├── Pool1.h5ad.tsv
    ├── vireo_annot.Pool1.h5ad
    └── Vireo_plots.pdf
```
per donor quantification matrix and additional metadata

## cellsnp
```
cellsnp
└── cellsnp_Pool1
    ├── cellSNP.base.vcf.gz
    ├── cellSNP.cells.vcf.gz
    ├── cellSNP.samples.tsv
    ├── cellSNP.tag.AD.mtx
    ├── cellSNP.tag.DP.mtx
    └── cellSNP.tag.OTH.mtx
```
genotypes called from single-cell data per (per droplet/cell)
## existing_cellsnp

## concordances
```
concordances
└── all_variants_description.tsv

concordances
├── all_variants_description.tsv
├── becoming_different_donor.png
├── becoming_doublet_donor.png
├── becoming_unassigned_donor.png
├── Pool1
│   ├── 1090095_1090095-donor3--each_cells_comparison_with_other_donor.tsv
│   ├── 1709635_1709635-donor5--each_cells_comparison_with_other_donor.tsv
│   ├── 2288590_2288590-donor6--each_cells_comparison_with_other_donor.tsv
│   ├── 2743244_2743244-donor7--each_cells_comparison_with_other_donor.tsv
│   ├── 2768849_2768849-donor4--each_cells_comparison_with_other_donor.tsv
│   ├── 2998395_2998395-donor2--each_cells_comparison_with_other_donor.tsv
│   ├── 3183427_3183427-donor0--each_cells_comparison_with_other_donor.tsv
│   ├── 3699286_3699286-donor1--each_cells_comparison_with_other_donor.tsv
│   ├── 4853673_4853673-donor9--each_cells_comparison_with_other_donor.tsv
│   ├── 5154993_5154993-donor8--each_cells_comparison_with_other_donor.tsv
│   ├── becoming_different_donor.png
│   ├── becoming_doublet_donor.png
│   ├── becoming_unassigned_donor.png
│   ├── cell_belongings.tsv
│   ├── cellSNP.cells.vcf.gz
│   ├── Pool1__joined_df_for_plots.tsv
│   ├── Pool1_subsampling_donor_swap_quantification.tsv
│   ├── Discordant_reads_becoming_different_donor_no0.png
│   ├── Discordant_reads_becoming_different_donor.png
│   ├── Discordant_reads_by_n_sites_becoming_different_donor_no0.png
│   ├── Discordant_reads_by_n_sites_becoming_different_donor.png
│   ├── discordant_sites_in_other_donors_noA2G.tsv
│   ├── Nr_discordant_uninformative_becoming_different_donor.png
│   ├── sites_becoming_different_donor_no0.png
│   ├── sites_becoming_different_donor.png
│   ├── sites_becoming_different_donor_probs.png
│   ├── sites_becoming_doublet_donor.png
│   ├── sites_becoming_unassigned_donor.png
│   ├── sites_vs_concordance.png
│   ├── stats_Pool1_gt_donor_assignments.csv
│   ├── sub_Pool1_Expected.vcf.gz
│   ├── sub_Pool1_GT_Matched.vcf.gz
│   ├── subplot_sites_vs_concordance.png
│   └── Total_reads_becoming_different_donor.png
├── Discordant_reads_becoming_different_donor_no0.png
├── Discordant_reads_becoming_different_donor.png
├── Discordant_reads_by_n_sites_becoming_different_donor_no0.png
├── Discordant_reads_by_n_sites_becoming_different_donor.png
├── joined_df_for_plots.tsv
├── Nr_discordant_uninformative_becoming_different_donor.png
├── sites_becoming_different_donor_no0.png
├── sites_becoming_different_donor.png
├── sites_becoming_different_donor_probs.png
├── sites_becoming_doublet_donor.png
├── sites_becoming_unassigned_donor.png
├── sites_vs_concordance.png
├── subplot_sites_vs_concordance.png
└── Total_reads_becoming_different_donor.png

```
statistics to describe how confident is the cell inferred genotype alignment with the reference genotypes.

## gtmatch
```
gtmatch/
├── assignments_all_pools.tsv
└── Pool1
    ├── Done.tmp
    ├── Expected_Withing_expected_Pool1.genome
    ├── GT_replace_PiHAT_Stats_File_Pool1.csv
    ├── InferedExpected_Expected_Infered_Pool1.genome
    ├── InferedGTMatched_Expected_Infered_Pool1.genome
    ├── InferedOnly_Withing_pool_Pool1.genome
    ├── PiHAT_Stats_File_Pool1.csv
    ├── Pool1_gt_donor_assignments.csv
    ├── pool_Pool1_panel_Pool1_Onek1K_gtcheck_donor_assignments.csv
    ├── pool_Pool1_panel_Pool1_Onek1K_gtcheck_score_table.csv
    └── stats_Pool1_gt_donor_assignments.csv
```
if genotypes are provided it contains the results of donors assigned by gtcheck.

## celltype_assignemt
```
celltype_assignemt/
├── All_Celltype_Assignments.tsv
├── azimuth
│   └── PBMC
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l1.mapping_score_umap.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l1.mapping_score_vln.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l1.ncells_by_type_barplot.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l1.prediction_score_umap.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l1.prediction_score_vln.pdf
│       ├── AZ_1.pre_QC_adata_Pool1_Pool1_celltype.l1.query_umap.pdf
│       └── AZ_1.pre_QC_adata_Pool1_Pool1_predicted_celltype_l1.tsv
├── celltypist
│   ├── COVID19_Immune_Landscape
│   │   └── Pool1
│   │       ├── Pool1___COVID19_Immune_Landscape___decision_matrix.csv
│   │       ├── Pool1___COVID19_Immune_Landscape___predicted_labels.csv
│   │       ├── Pool1___COVID19_Immune_Landscape___probability_matrix.csv
│   │       ├── Pool1_majority_voting.pdf
│   │       ├── Pool1_over_clustering.pdf
│   │       └── Pool1_predicted_labels.pdf
│   ├── Immune_All_High
│   │   └── Pool1
│   │       ├── Pool1___Immune_All_High___decision_matrix.csv
│   │       ├── Pool1___Immune_All_High___predicted_labels.csv
│   │       ├── Pool1___Immune_All_High___probability_matrix.csv
│   │       ├── Pool1_majority_voting.pdf
│   │       ├── Pool1_over_clustering.pdf
│   │       └── Pool1_predicted_labels.pdf
│   └── Immune_All_Low
│       └── Pool1
│           ├── Pool1___Immune_All_Low___decision_matrix.csv
│           ├── Pool1___Immune_All_Low___predicted_labels.csv
│           ├── Pool1___Immune_All_Low___probability_matrix.csv
│           ├── Pool1_majority_voting.pdf
│           ├── Pool1_over_clustering.pdf
│           └── Pool1_predicted_labels.pdf
├── donor_celltype_report.tsv
├── scpred
│   ├── AZ_1.pre_QC_adata_Pool1_AZ_1.pre_QC_adata_Pool1__scpred_prediction.tsv
│   └── AZ_1.pre_QC_adata_Pool1_hier_scpred.RDS
└── tranche_celltype_report.tsv
```
- **[All_Celltype_Assignments.tsv] combination of results per barcode/droplet/cell
- **[donor_celltype_report.tsv] per donor summarised cell numbers
- **[tranche_celltype_report.tsv] per tranch summarised cell numbers
cell type assignments by different methods and models:
- **[scored]
- **[azimuth]
- **[celltypist]
- Immune_All_Low
- COVID19_Immune_Landscape
- Immune_All_High

## clustering_and_integration
```
clustering_and_integration/
├── normalize=total_count.vars_to_regress=none
│   ├── adatametadata.tsv.gz
│   ├── adatanormalized.h5ad
│   ├── adatanormalized_pcacounts.h5ad
│   ├── adatanormalized_pca.h5ad
│   ├── adatanormalized_pcaknee.tsv
│   ├── adatapcs.tsv.gz
│   ├── donor_level_anndata_QCfiltered
│   │   └── Pool1___sample_QCd_adata.h5ad
│   ├── plots
│   ├── reduced_dims-null-bbknn.batch=experiment_id.n_pcs=20
│   │   ├── cluster.number_neighbors=-1.method=leiden.resolution=0.1
│   │   │   ├── clustering_0.1clustered.h5ad
│   │   │   ├── clustering_0.1clustered.tsv.gz
│   │   │   ├── dotplot_sampleclustering_0.1clustered_ncells0.pdf
│   │   │   ├── dotplot_sampleclustering_0.1clustered_ncellsless5.pdf
│   │   │   ├── dotplot_sampleclustering_0.1clustered.pdf
│   │   │   ├── plots
│   │   │   ├── sccaf
│   │   │   └── validate_resolution
│   │   ├── cluster.number_neighbors=-1.method=leiden.resolution=0.5
│   │   ├── cluster.number_neighbors=-1.method=leiden.resolution=1.0
│   │   ├── cluster.number_neighbors=-1.method=leiden.resolution=5.0
│   │   ├── outfile_adatabbknn.h5ad
│   │   ├── plots
│   │   ├── reduced_dims.tsv.gz
│   │   ├── resolution_tuningmerged_model_report.tsv.gz
│   │   ├── resolution_tuningmerged_test_result.tsv.gz
│   │   └── umap_gather_out.h5ad
│   ├── reduced_dims-null-harmony.n_pcs=20.variables=experiment_id.thetas=1.0
│   │   ├── cluster.number_neighbors=15.method=leiden.resolution=0.1
│   │   ├── cluster.number_neighbors=15.method=leiden.resolution=0.5
│   │   ├── cluster.number_neighbors=15.method=leiden.resolution=1.0
│   │   ├── cluster.number_neighbors=15.method=leiden.resolution=5.0
│   │   ├── plots
│   │   ├── reduced_dims.tsv.gz
│   │   ├── resolution_tuningmerged_model_report.tsv.gz
│   │   ├── resolution_tuningmerged_test_result.tsv.gz
│   │   └── umap_gather_out.h5ad
│   └── reduced_dims-null-pca.n_pcs=20
│       ├── clustering_and_integration
│       │   └── plots
│       └── reduced_dims.tsv.gz
└── plots

```
integrated and clustered data with statistics and plots describing integration and clustering performance

normalize=total_count.vars_to_regress=none (should it be renamed)
plots

## citeseq
```
citeseq/
└── DSB
    └── Pool1
        └── CITE__Pool1
```
- **[DSB] is there only one folder in citeseq? DSB has folders for each pool
contains DSB background removed protein counts if citeseq is in quantification matrix
## handover
- **[Summary_plots]
- **[Donor_Quantification]
- **[Donor_Quantification_summary]
- **[merged_h5ad]

```
handover/
├── Donor_Quantification
├── Donor_Quantification_summary
├── merged_h5ad
└── Summary_plots
└── UMAPs
```

## Pipeline Info

## YASCP Inputs

